### PR TITLE
Add a scheduler running the system functions once at each frame.

### PR DESCRIPTION
**Pull request by catageek**
Each system registered has a dedicated thread that executes its 2 main functions every 1/60 second.

When not running the system functions, the scheduler assigns pending tasks to threads. The tasks can be split in small blocks to allow collaborative behaviour (the NetworkController runs entirely asynchronusly using the scheduler). The keyboard and mouse controllers should us asynchronous tasks only.

For illustration, a FakeSystem has been added, and its 2 functions implemented to print a message each time they are executed.

The scheduler uses a constant number of threads and limits the number active threads. The implementation is still a work in progress, but the interface should not change.

### CODE REVIEW

File: include/System.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+         *
+         * This function is called once every frame. It is the only
+         * function that can write data. This function is in the critical
+         * path, job done here must be simple.
+         *
+         * If event handling need some batch processing, a task list must be
+         * prepared and stored temporarily to be retrieved by RunBatch().
+         *
+         */
+        virtual void HandleEvents(const std::chrono::time_point<std::chrono::steady_clock, frame_unit>& timepoint) = 0;
```
Comment (adam4813) [inline include/System.h]: So, this would be equivalent to the Update method from Sigma. in a general sense.

Comment (catageek) [inline include/System.h]: HandleEvents() followed by RunBatch() is equivalent to Update(), but split in 2 parts.

**Comment by adam4813** [comment]
I'll give this a windows test tomorrow hopefully and we can merge it in assuming it'll work with the design.

**Comment by adam4813** [comment]
Ok...so for some reason on Visual Studio steady_clock::now() calls system_clock::now() which can't be compared to a steady_clock such as in the line 32.

Also Visual Studio doesn't have thread_local, but it does have __declspec(thread). Unfortunately __declspec(thread) doesn't work for non-POD.

**Comment by catageek** [comment]
I removed thread_local and use local variables instead.

For the clock issue, I need to implement something different, but it does not prevent the merge since it is an internal issue.

EDIT : We will use the GLFW clock function once GLFW and this PR will be merged

File: include/TrillekScheduler.h [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        virtual ~TaskRequestBase() {};
+
+        bool operator<(const TaskRequestBase& tqe) const {
+            return this->timestamp < tqe.timestamp;
+        }
+
+        virtual void RunTask() = 0;
+
+        bool IsNow() const {
+            return timestamp < steady_clock::now();
```
Comment (adam4813) [inline include/TrillekScheduler.h]: This line fails to compile on visual studio because steady_clock::now() calls system_clock::now() which for some reason can't be compared even though steady_clock is a subclass of system_clock.

File: include/TrillekScheduler.h [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        void Reschedule(frame_unit&& delay) {
+            timestamp = steady_clock::now() + delay;
```
Comment (adam4813) [inline include/TrillekScheduler.h]: Same issue as above, but with attempting to add a system_clock to a steady_clock.

**Comment by catageek** [comment]
I forgot to replace the calls of now() in header. It is fixed.

**Comment by catageek** [comment]
We have now a clean program termination.

**Comment by catageek** [comment]
Any reference to steady_clock will be replaced by system_clock in time_point definition, since the only value that time_point wants to access is the epoch of the clock. I hope that MSVC will be happy.

Also, I will try to cut lines to be more readable...

**Comment by milesrout** [comment]
@adam4813 Merge this.

**Comment by adam4813** [comment]
Why, when it breaks all visual c++ compiling?

I am still not certain as to what multithreading at this stage gets us. Unless we are utilizing out of order logic processing for each system on separate cores, threading won't grant us any major benefits. 

Tomorrow I will try and sit down with @catageek and determine a way to get this to compile via visual c++ as we will be excluding a chunk of potential modders that are windows lovers such as myself.

**Comment by milesrout** [comment]
If it breaks then fix it, @catageek can't compile on Windows as you have known for weeks.

**Comment by milesrout** [comment]
If you don't build around being thread-safe from the start it will be very difficult to add it in later. You know that, I don't know why I'm having to tell you this.

**Comment by catageek** [comment]
Ready to be merged to the current develop branch and VS2013-compatible. The window control is now transferred to the graphic system thread (OpenGL is not multithread-friendly) and returned to the main thread when closing.

Theoretically, all the graphic stuff, including the window creation, should be moved to the graphic system to have a trivial main().

I have created an ThreadInit function in the SystemBase interface that is executed once the system has its own thread. It is used to make the context of the window current to the thread. All the Initialize() function may be moved to it.

Basically, nothing should be initialized in main except the scheduler with the list of systems to start, and the controllers tasks already queued.

I left things in this state because I am not sure of the status of the "transform" and "resource" instances created in main, and if the can/must be moved.

**Comment by milesrout** [comment]
@catageek VS2013 with the free CTP supports rvalue references, you don't have to remove them. 

and.. uhh.. did you just make a whole lot of changes in that last commit that had already been made elsewhere in the repo? That's what a merge is for. 

Also the endless `typedef`s make the code very difficult to read.

**Comment by catageek** [comment]
@milesrout I spent hours to make this thing built by VS without errors, I leave the improvements to people who have VS and don't have to wait 5 minutes to get the result of the build like in appveyor.

I don't think I retyped some changes made in other commits. I merged the current develop branch to work on the latter version.

And typedefs are here to respect the holy limit of 80 chars on each line.

**Comment by milesrout** [comment]
If the code requires typedefs to respect 80 chars per line then it's too
complicated and should be refactored into smaller functions. If you can
read it on one screen then it's too complicated and should be refactored
into smaller functions. That's part of the benefit of auto, in fact.

Nice work, though, nice work.
